File: robo/api/risk.py
from fastapi import APIRouter, HTTPException, Path

from schemas.risk_schemas import (
    QuestionnaireGetResponse,
    QuestionResponse,
    AnswerOption,
    QuestionnaireResponse,
    ScoreWithCategoryResponse,
    QuestionScore,
)
from services.risk_service import load_questions

router = APIRouter()


@router.get("/questions/{risk_type}", response_model=QuestionnaireGetResponse)
def get_questions(
    risk_type: str = Path(..., regex="^(tolerance|capacity)$", description="Risk type: 'tolerance' or 'capacity'")
):
    try:
        questions = load_questions(risk_type)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    response_questions = [
        QuestionResponse(
            questionText=q.questionText,
            answers=[AnswerOption(answerText=a.answerText) for a in q.answers],
        )
        for q in questions
    ]

    return {"type": risk_type, "questions": response_questions}


@router.post("/score/{risk_type}", response_model=ScoreWithCategoryResponse)
def post_score(
    data: QuestionnaireResponse,
    risk_type: str = Path(..., regex="^(tolerance|capacity)$", description="Risk type: 'tolerance' or 'capacity'"),
):
    try:
        questions = load_questions(risk_type)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    num_questions = len(questions)
    total_score = 0
    per_question_scores = []

    # Validate answers and compute scores
    for ans in data.answers:
        if not (0 <= ans.questionIndex < num_questions):
            raise HTTPException(status_code=400, detail=f"Invalid questionIndex: {ans.questionIndex}")

        question = questions[ans.questionIndex]

        if not (0 <= ans.answerIndex < len(question.answers)):
            raise HTTPException(
                status_code=400,
                detail=f"Invalid answerIndex: {ans.answerIndex} for question {ans.questionIndex}",
            )

        answer_score = question.answers[ans.answerIndex].score * question.weight
        total_score += answer_score
        per_question_scores.append(QuestionScore(questionText=question.questionText, score=answer_score))

    # Calculate max possible score
    max_total_score = sum(max(ans.score for ans in q.answers) * q.weight for q in questions)
    category_threshold = max_total_score / 3

    if total_score <= category_threshold:
        category = "Low"
    elif total_score <= 2 * category_threshold:
        category = "Medium"
    else:
        category = "High"

    return ScoreWithCategoryResponse(totalScore=total_score, perQuestionScores=per_question_scores, category=category)

File: client/lib/riskProfiling/risk.ts
// Types
export interface AnswerOption {
  answerText: string;
}

export interface QuestionResponse {
  questionText: string;
  answers: AnswerOption[];
}

export interface QuestionnaireGetResponse {
  type: string;
  questions: QuestionResponse[];
}

export interface UserAnswer {
  questionIndex: number;
  answerIndex: number;
}

export interface QuestionnaireResponse {
  answers: UserAnswer[];
}

export interface QuestionScore {
  questionText: string;
  score: number;
}

export interface ScoreWithCategoryResponse {
  totalScore: number;
  perQuestionScores: QuestionScore[];
  category: 'Low' | 'Medium' | 'High';
  bucket?: string; 
}

export type RiskType = 'tolerance' | 'capacity';


const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8000/risk';


class RiskProfilingAPI {
  private baseUrl: string;

  constructor(baseUrl: string = API_BASE_URL) {
    this.baseUrl = baseUrl;
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
    }
    return response.json();
  }

  async getQuestions(riskType: RiskType): Promise<QuestionnaireGetResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/questions/${riskType}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      return this.handleResponse<QuestionnaireGetResponse>(response);
    } catch (error) {
      console.error(`Error fetching ${riskType} questions:`, error);
      throw new Error(`Failed to fetch ${riskType} questions. Please try again.`);
    }
  }

  async submitAnswers(
    riskType: RiskType, 
    answers: UserAnswer[]
  ): Promise<ScoreWithCategoryResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/score/${riskType}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ answers }),
      });

      return this.handleResponse<ScoreWithCategoryResponse>(response);
    } catch (error) {
      console.error(`Error submitting ${riskType} answers:`, error);
      throw new Error(`Failed to submit ${riskType} answers. Please try again.`);
    }
  }

  async getAllQuestions(): Promise<{
    tolerance: QuestionnaireGetResponse;
    capacity: QuestionnaireGetResponse;
  }> {
    try {
      const [tolerance, capacity] = await Promise.all([
        this.getQuestions('tolerance'),
        this.getQuestions('capacity'),
      ]);

      return { tolerance, capacity };
    } catch (error) {
      console.error('Error fetching all questions:', error);
      throw new Error('Failed to fetch assessment questions. Please try again.');
    }
  }


  async submitAllAnswers(data: {
    toleranceAnswers: UserAnswer[];
    capacityAnswers: UserAnswer[];
  }): Promise<{
    tolerance: ScoreWithCategoryResponse;
    capacity: ScoreWithCategoryResponse;
  }> {
    try {
      const [tolerance, capacity] = await Promise.all([
        this.submitAnswers('tolerance', data.toleranceAnswers),
        this.submitAnswers('capacity', data.capacityAnswers),
      ]);

      return { tolerance, capacity };
    } catch (error) {
      console.error('Error submitting all answers:', error);
      throw new Error('Failed to submit assessment answers. Please try again.');
    }
  }
}

export const riskProfilingAPI = new RiskProfilingAPI();

export interface UseQuestionsResult {
  data: QuestionnaireGetResponse | null;
  loading: boolean;
  error: string | null;
  refetch: () => Promise<void>;
}

export interface UseAllQuestionsResult {
  data: {
    tolerance: QuestionnaireGetResponse;
    capacity: QuestionnaireGetResponse;
  } | null;
  loading: boolean;
  error: string | null;
  refetch: () => Promise<void>;
}

export interface UseSubmitAnswersResult {
  submitAnswers: (riskType: RiskType, answers: UserAnswer[]) => Promise<ScoreWithCategoryResponse>;
  submitting: boolean;
  error: string | null;
}

export interface UseSubmitAllAnswersResult {
  submitAllAnswers: (data: {
    toleranceAnswers: UserAnswer[];
    capacityAnswers: UserAnswer[];
  }) => Promise<{
    tolerance: ScoreWithCategoryResponse;
    capacity: ScoreWithCategoryResponse;
  }>;
  submitting: boolean;
  error: string | null;
}

export const getRiskCategoryColor = (category: string): string => {
  switch (category.toLowerCase()) {
    case 'low':
      return 'text-green-600 bg-green-100';
    case 'medium':
      return 'text-yellow-600 bg-yellow-100';
    case 'high':
      return 'text-red-600 bg-red-100';
    default:
      return 'text-gray-600 bg-gray-100';
  }
};

export const getRiskCategoryIcon = (category: string): string => {
  switch (category.toLowerCase()) {
    case 'low':
      return '🛡️';
    case 'medium':
      return '⚖️';
    case 'high':
      return '🚀';
    default:
      return '❓';
  }
};

export const formatRiskDescription = (category: string): string => {
  switch (category.toLowerCase()) {
    case 'low':
      return 'Conservative approach with focus on capital preservation';
    case 'medium':
      return 'Balanced approach with moderate risk and return expectations';
    case 'high':
      return 'Aggressive approach with higher risk tolerance for potential gains';
    default:
      return 'Risk profile assessment needed';
  }
};

// Validation helpers
export const validateAnswers = (
  answers: UserAnswer[], 
  totalQuestions: number
): { isValid: boolean; errors: string[] } => {
  const errors: string[] = [];

  if (answers.length !== totalQuestions) {
    errors.push(`Expected ${totalQuestions} answers, but received ${answers.length}`);
  }

  answers.forEach((answer, index) => {
    if (answer.questionIndex < 0 || answer.questionIndex >= totalQuestions) {
      errors.push(`Invalid question index ${answer.questionIndex} at position ${index}`);
    }
    if (answer.answerIndex < 0) {
      errors.push(`Invalid answer index ${answer.answerIndex} at position ${index}`);
    }
  });

  const questionIndices = answers.map(a => a.questionIndex);
  const uniqueIndices = new Set(questionIndices);
  if (uniqueIndices.size !== questionIndices.length) {
    errors.push('Duplicate question indices found');
  }

  return {
    isValid: errors.length === 0,
    errors
  };
};

export default riskProfilingAPI;